File: AIReader/AIReader/Domain/Formats/DSLDictionaryReader.swift
import Foundation

/// ABBYY Lingvo's DSL. Headwords start at column zero; the lines indented
/// under them are the article, marked up with `[tags]` this reader strips.
enum DSLDictionaryReader {
    static func read(
        _ source: DictionarySource,
        entry: (DictionaryImportEntry) throws -> Void
    ) throws -> DictionaryImportInfo {
        var info = DictionaryImportInfo(name: source.defaultName)
        var headwords: [String] = []
        var senses: [String] = []

        func flush() throws {
            defer { headwords = []; senses = [] }
            guard !headwords.isEmpty, !senses.isEmpty else { return }
            for headword in headwords {
                try entry(
                    DictionaryImportEntry(headword: headword, partOfSpeech: nil, senses: senses)
                )
            }
        }

        for line in try TextFile.lines(at: source.main) {
            if let directive = directive(in: line) {
                switch directive.0 {
                case "#NAME": info.name = directive.1
                case "#INDEX_LANGUAGE": info.targetLanguage = LanguageName.code(directive.1)
                case "#CONTENTS_LANGUAGE": info.definitionLanguage = LanguageName.code(directive.1)
                default: break
                }
                continue
            }

            let isIndented = line.first == "\t" || line.first == " "
            let text = strip(String(line)).trimmingCharacters(in: .whitespaces)

            if isIndented {
                if !text.isEmpty { senses.append(text) }
            } else {
                // A run of headwords shares the article indented below it.
                if !senses.isEmpty { try flush() }
                if !text.isEmpty { headwords.append(text) }
            }
        }
        try flush()
        return info
    }

    /// `#NAME "Big Dictionary"` split into its parts.
    private static func directive(in line: Substring) -> (String, String)? {
        guard line.hasPrefix("#") else { return nil }
        let parts = line.split(separator: " ", maxSplits: 1)
        guard let key = parts.first else { return nil }
        let value = parts.count > 1 ? String(parts[1]) : ""
        return (String(key).uppercased(), value.trimmingCharacters(in: CharacterSet(charactersIn: " \"\r")))
    }

    /// Removes DSL markup, keeping the words inside it.
    ///
    /// `[...]` are formatting tags, `{{...}}` are comments, and a backslash
    /// escapes the character after it.
    static func strip(_ line: String) -> String {
        var output = ""
        var depth = 0
        var index = line.startIndex

        while index < line.endIndex {
            let character = line[index]
            let next = line.index(after: index)

            if character == "\\", next < line.endIndex {
                output.append(line[next])
                index = line.index(after: next)
                continue
            }
            if character == "{", next < line.endIndex, line[next] == "{" {
                if let end = line.range(of: "}}", range: next..<line.endIndex) {
                    index = end.upperBound
                    continue
                }
            }
            switch character {
            case "[": depth += 1
            case "]": depth = max(0, depth - 1)
            case "{", "}": break  // headword parts that are shown but not indexed
            default: if depth == 0 { output.append(character) }
            }
            index = next
        }
        return output.replacingOccurrences(of: "\r", with: "")
    }
}

File: AIReaderKindle/src/Domain/Formats/DSLDictionaryReader.cpp
#include "DSLDictionaryReader.hpp"

#include "../../Support/Text.hpp"
#include "../../Support/TextFile.hpp"

namespace DSLDictionaryReader {

/// `#NAME "Big Dictionary"` split into its parts.
static std::optional<std::pair<std::string, std::string>> directive(const std::string& line) {
    if (!Text::startsWith(line, "#")) return std::nullopt;
    auto space = line.find(' ');
    std::string key = Text::lower(space == std::string::npos ? line : line.substr(0, space));
    std::string value = space == std::string::npos ? "" : line.substr(space + 1);
    while (!value.empty() && (value.front() == ' ' || value.front() == '"')) value.erase(0, 1);
    while (!value.empty() && (value.back() == ' ' || value.back() == '"' || value.back() == '\r')) value.pop_back();
    return std::make_pair(key, value);
}

std::string strip(const std::string& line) {
    std::string output;
    int depth = 0;
    for (size_t i = 0; i < line.size(); ++i) {
        char c = line[i];
        // A backslash escapes the character after it.
        if (c == '\\' && i + 1 < line.size()) {
            output += line[++i];
            continue;
        }
        // `{{...}}` are comments.
        if (c == '{' && i + 1 < line.size() && line[i + 1] == '{') {
            auto end = line.find("}}", i + 2);
            if (end != std::string::npos) {
                i = end + 1;
                continue;
            }
        }
        switch (c) {
        case '[': ++depth; break;
        case ']': if (depth) --depth; break;
        case '{': case '}': break;  // headword parts that are shown but not indexed
        case '\r': break;
        default: if (depth == 0) output += c;
        }
    }
    return output;
}

std::optional<DictionaryImportInfo> read(const DictionarySource& source, const DictionaryEntrySink& entry) {
    auto lines = TextFile::lines(source.main);
    if (!lines) return std::nullopt;

    DictionaryImportInfo info{source.defaultName(), "", ""};
    std::vector<std::string> headwords;
    std::vector<std::string> senses;
    auto flush = [&] {
        if (!headwords.empty() && !senses.empty()) {
            for (const auto& headword : headwords) entry(DictionaryImportEntry{headword, "", senses});
        }
        headwords.clear();
        senses.clear();
    };

    for (const auto& line : *lines) {
        if (auto found = directive(line)) {
            if (found->first == "#name") info.name = found->second;
            else if (found->first == "#index_language") info.targetLanguage = LanguageName::code(found->second);
            else if (found->first == "#contents_language") info.definitionLanguage = LanguageName::code(found->second);
            continue;
        }
        bool indented = !line.empty() && (line[0] == '\t' || line[0] == ' ');
        std::string text = Text::trim(strip(line));
        if (indented) {
            if (!text.empty()) senses.push_back(text);
        } else {
            // A run of headwords shares the article indented below it.
            if (!senses.empty()) flush();
            if (!text.empty()) headwords.push_back(text);
        }
    }
    flush();
    return info;
}

}  // namespace DSLDictionaryReader
